presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedTruncatedHistogram.java

```diff
@@ -0,0 +1,318 @@
+                counts.add(hashPositions.get(hashPosition), count);
+                return;
+            }
+
+            // increment position and mask to handle wrap around
+            hashPosition = (hashPosition + 1) & mask;
+        }
+
+        addNewGroup(hashPosition, position, block, count);
+    }
+
+    private void addNewGroup(int hashPosition, int position, Block block, long count)
+    {
+        hashPositions.set(hashPosition, values.getPositionCount());
+        counts.set(values.getPositionCount(), count);
+        type.appendTo(block, position, values);
+
+        compactIfNeeded();
+
+        rehashIfNeeded();
+
+        if (getEstimatedSize() > FOUR_MEGABYTES) {
+            throw exceededLocalLimit(new DataSize(4, MEGABYTE));
+        }
+    }
+
+    private void rehashIfNeeded()
+    {
+        // increase capacity, if necessary
+        if (values.getPositionCount() >= maxFill) {
+            int size = maxFill * 2;
+            int hashSize = arraySize(size + 1, FILL_RATIO);
+            mask = hashSize - 1;
+            rehash();
+        }
+    }
+
+    private void rehash()
+    {
+        int hashSize = mask + 1;
+        IntBigArray newHashPositions = new IntBigArray(-1);
+        newHashPositions.ensureCapacity(hashSize);
+
+        for (int i = 0; i < values.getPositionCount(); i++) {
+            // find an empty slot for the address
+            int hashPosition = getHashPosition(com.facebook.presto.type.TypeUtils.hashPosition(type, values, i), mask);
+            while (newHashPositions.get(hashPosition) != -1) {
+                hashPosition = (hashPosition + 1) & mask;
+            }
+
+            // record the mapping
+            newHashPositions.set(hashPosition, i);
+        }
+
+        maxFill = calculateMaxFill(hashSize);
+        hashPositions = newHashPositions;
+
+        this.counts.ensureCapacity(maxFill);
+    }
+
+    /**
+     * Call {@link #compact()} if over {@link #COMPACT_RATIO} of the value/count pairs exceeds {@link #compactLimit}.
+     */
+    private void compactIfNeeded()
+    {
+        if (values.getPositionCount() * COMPACT_RATIO > compactLimit) {
+            compact();
+        }
+    }
+
+    /**
+     * Ensure there are no more than {@link #compactLimit} value/counts pairs.
+     * This works by evicting the smallest counts and overflowing the counts to the smallest retained counts, increasing
+     * them all to the same value (rounding up).
+     * The end result is that the estimated counts are an upper bound to the true counts.
+     * In addition, the over-estimation is less than the minimum retained count.
+     * This is because every overflow increases the minimum retained count by the most.
+     */
+    private void compact()
+    {
+        if (values.getPositionCount() <= compactLimit) {
+            return;
+        }
+
+        int truncatedNum = values.getPositionCount() - compactLimit;
+        assert 0 < truncatedNum;
```

Comment:
We don't use `assert`, use `checkState` or `checkArgument` instead, as appropriate